keylime-agent/src/payloads.rs

```diff
@@ -0,0 +1,481 @@
+            Ok(())
+        }
+        Err(e) => Err(Error::Other(format!(
+            "{:?} failed during run: {}",
+            &script_path, e
+        ))),
+    }
+}
+
+// checks if keylime-agent.conf indicates the payload should be unzipped, and does so if needed.
+// the input string is the directory where the unzipped file(s) should be stored.
+pub(crate) fn optional_unzip_payload(
+    unzipped: &Path,
+    config: &config::KeylimeConfig,
+) -> Result<()> {
+    if config.agent.extract_payload_zip {
+        let zipped_payload = &config.agent.dec_payload_file;
+        let zipped_payload_path = unzipped.join(zipped_payload);
+
+        info!("Unzipping payload {} to {:?}", &zipped_payload, unzipped);
+
+        let mut source = fs::File::open(&zipped_payload_path)?;
+        uncompress_archive(&mut source, unzipped, Ownership::Ignore)?;
+    }
+
+    Ok(())
+}
+
+pub(crate) async fn run_encrypted_payload(
+    symm_key: SymmKey,
+    payload: EncryptedData,
+    config: &config::KeylimeConfig,
+    mount: &Path,
+) -> Result<()> {
+    let dec_payload = decrypt_payload(&symm_key, payload)?;
+
+    let (unzipped, dec_payload_path, key_path) =
+        setup_unzipped(config, mount)?;
+
+    write_out_key_and_payload(
+        &dec_payload,
+        &dec_payload_path,
+        &symm_key,
+        &key_path,
+    )?;
+
+    optional_unzip_payload(&unzipped, config)?;
+    // there may also be also a separate init script
+    match config.agent.payload_script.as_str() {
+        "" => {
+            info!("No payload script specified, skipping");
+        }
+        script => {
+            info!("Payload init script indicated: {}", script);
+            run(&unzipped, script, config.agent.uuid.as_str()).await?;
+        }
+    }
+
+    // Set execution permission for listed revocation actions
+    let action_file = unzipped.join("action_list");
+
+    if action_file.exists() {
+        let action_data = std::fs::read_to_string(&action_file)
+            .expect("unable to read action_list");
+
+        action_data
+            .split('\n')
+            .filter(|&script| !script.is_empty())
+            .map(|script| script.trim())
+            .map(|script| unzipped.join(script))
+            .filter(|script| script.exists())
+            .try_for_each(|script| {
+                if fs::set_permissions(
+                    &script,
+                    fs::Permissions::from_mode(0o700),
+                )
+                .is_err()
+                {
+                    error!(
+                        "Could not set permission for action {}",
+                        script.display()
+                    );
+                    Err(Error::Permission)
+                } else {
+                    info!("Permission set for action: {}", script.display());
+                    Ok(())
+                }
+            })?
+    }
+
+    Ok(())
+}
+
+pub(crate) async fn worker(
+    config: config::KeylimeConfig,
+    mount: PathBuf,
```

Comment:
I understand that the arguments of `worker` needs to be converted to an owned value before being sent to another thread, though it is a bit confusing to use owned value (`PathBuf`) and shared reference (`&Path`) in an intermixed manner. Maybe we could use `AsRef<Path>`?

```suggestion
    mount: impl AsRef<Path>,
```

Reply:
Done! Thanks!